x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -252,11 +249,39 @@ public IndexMetadata getSnapshotIndexMetaData(RepositoryData repositoryData, Sna
         return imdBuilder.build();
     }
 
+    /**
+     * Dummy request key for deduplicating all remote cluster state requests via {@link #getRemoteStateDeduplicator}.
+     */
+    private static final Object RESULT_KEY = new Object();
+
+    /**
+     * Fetches the remote cluster state from the leader cluster, deduplicating the result if there is already an ongoing call
+     * to this method.
+     */
+    private void getRemoteState(ActionListener<ClusterState> listener) {
+        // We only allow a single remote cluster state requests at a time. The callbacks to the cluster state responses run on the
+        // transport thread and can safely assume they are fast enough so that this does not lead to seeing substantially outdated
+        // remote states as a result of a hot loop calling this method ever.
+        getRemoteStateDeduplicator.executeOnce(
+            RESULT_KEY,
+            listener,
+            (ignored, l) -> getRemoteClusterClient().admin()
+                .cluster()
+                .prepareState()
+                .clear()
+                .setMetadata(true)
+                .setNodes(true)
+                .setMasterNodeTimeout(TimeValue.MAX_VALUE)
+                // fork to the snapshot meta pool because the context expects to run on it and asserts that it does
```

Comment:
nit: this comment is lost, it should be next to the ThreadedActionListener in one of the callers.